frontend/cli/cmd_lsp.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"connectrpc.com/connect"
+	"golang.org/x/sync/errgroup"
+
+	buildenginepb "github.com/block/ftl/backend/protos/xyz/block/ftl/buildengine/v1"
+	enginepbconnect "github.com/block/ftl/backend/protos/xyz/block/ftl/buildengine/v1/buildenginepbconnect"
+	"github.com/block/ftl/internal/log"
+	"github.com/block/ftl/internal/lsp"
+	"github.com/block/ftl/internal/rpc"
+)
+
+type lspCmd struct {
+	BuildUpdatesEndpoint *url.URL `help:"Build updates endpoint." default:"http://127.0.0.1:8900" env:"FTL_BUILD_UPDATES_ENDPOINT"`
+	languageServer       *lsp.Server
+}
+
+func (l *lspCmd) Run(ctx context.Context) error {
+	logger := log.FromContext(ctx).Scope("lsp")
+	logger.Infof("Starting LSP server and listening for updates on %s", l.BuildUpdatesEndpoint)
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	l.languageServer = lsp.NewServer(ctx)
+	g.Go(func() error {
+		return l.languageServer.Run()
+	})
+
+	client := rpc.Dial(enginepbconnect.NewBuildEngineServiceClient, l.BuildUpdatesEndpoint.String(), log.Error)
+	g.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				logger.Debugf("Connecting to build updates service")
+				if err := l.streamBuildEvents(ctx, client); err != nil {
+					logger.Debugf("Failed to connect to build updates service: %v", err)
```

Comment:
Similar logic to the comment about initial state in the stream, theres bad states that could arise here:
eg: errors were present in buildengine, lost connection / build engine dies, when we reconnect those errors may not be present anymore but the LSP wont be notified. Maybe we need to reset lsp state here.